### PR TITLE
Improvements to Ex. 21 extraction model

### LINKED ISSUE #68: Improve performance of Ex. 21 extraction model
Now that we have a validation framework for Ex. 21 extraction, try these simple improvements and re-evaluate performance.

```[tasklist]
### Tasks
- [x] Investigate CorpWatch dataset to see if we can use anything there for validation
- [x] Get to 50 validation dataframes, high overlap with training data
- [x] Strip company name parts (LLC, Co, etc.) from names before doing similarity comparison
- [x] Retrain LayoutLM with more training data
- [x] Look at the diff between computed and validation dataframes
- [x] Handle null values in pecision and recall metrics
- [ ] Log metric with the number of docs in the training set, or log the training set metadata
- [x] Label more training set documents in Label Studio (get to 100 docs)
- [ ] Add an x and y threshold for distinct subsidiary cutoffs
```

### PR DESCRIPTION
**Pull request by katie-lamb**
# Overview

This is a pretty small PR to improve the Ex. 21 model. Since the Dagster integration, I think some of the larger structural changes will be easier to make once they're branched off the `error_handling_improvements` branch, so I think we should try and merge these little changes and then split the rest of the Ex. 21 model improvements into a separate PR.

Closes #68 .

What problem does this address?

What did you change in this PR?

* handle null values in precision and recall + jaccard similarity metric
* add a test for the cleaning done before these metrics are run
* handle change in filenames when creating Label Studio inputs and training data
* one off fixes to PDF rendering

# Testing

Ran Ex. 21 validation with MLflow to ensure expected results.

# To-do list

```[tasklist]
- [x] Update relevant documentation - like comments, docstrings, README, release notes, etc.
- [x] Review the PR yourself and call out any questions or issues you have
```

### CODE REVIEW

File: src/mozilla_sec_eia/library/validation_helpers.py [hunk 1]
```diff
@@ -65,17 +70,56 @@ def jaccard_similarity(
...
+            "Ownership Percentage": "own_per",
+        }
+    )
+    validation_df["own_per"] = validation_df["own_per"].astype(str)
+    validation_df["filename"] = validation_df["id"].apply(get_metadata_filename)
+    validation_df = clean_extracted_df(validation_df)
+    return validation_df
+
+
+def strip_down_company_names(ser: pd.Series) -> pd.Series:
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/library/validation_helpers.py:96]: Before computing comparison metrics between the expected validation subsidiary column and the extracted subsidiary column, I strip off company name suffixes (i.e. "LLC") to make the comparison more accurate. The suffixes are kept on the actual subsidiary columns that are extracted into tables, this is just used for validation.

File: src/mozilla_sec_eia/library/validation_helpers.py [hunk 2]
```diff
@@ -65,17 +70,56 @@ def jaccard_similarity(
...
+    reg = r"\b(?:" + "|".join(re.escape(word) for word in terms_list) + r")\b"
+    ser = ser.replace(reg, "", regex=True)
+    ser = ser.str.strip()
+    # strip commas or other special chars that might be at the end of the name
+    ser = ser.str.replace(r"[^[^\w&\s]+|[^\w&\s]+$]+", "", regex=True)
+    ser = ser.str.strip()
+    return ser
+
+
+def fill_nulls_for_comparison(ser: pd.Series):
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/library/validation_helpers.py:119]: This is also just used for validation. The comparison metrics weren't handling nulls before and were marking them as not equal.

File: src/mozilla_sec_eia/library/validation_helpers.py [hunk 3]
```diff
@@ -65,17 +70,56 @@ def jaccard_similarity(
...
+
+def strip_down_company_names(ser: pd.Series) -> pd.Series:
+    """Strip LLC and other company name suffixes from a column.
+
+    Used to compare subsidiary name columns during validation.
+    """
+    # this JSON is taken from PUDL package data (used for CompanyNameCleaner)
+    json_source = (
+        resources.files("mozilla_sec_eia.package_data") / "us_legal_forms.json"
+    )
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/library/validation_helpers.py:104]: This JSON is used in PUDL to clean company names before record linkage

File: src/mozilla_sec_eia/models/sec10k/ex_21/create_labeled_dataset.py [hunk 4]
```diff
@@ -157,6 +158,9 @@ def format_label_studio_output(
         with Path.open(json_file_path) as j:
             doc_dict = json.loads(j.read())
             filename = doc_dict["task"]["data"]["ocr"].split("/")[-1].split(".")[0]
+            # check if old local naming schema is being used
+            if len(filename.split("-")) == 6:
+                filename = "-".join(filename.split("-")[2:])
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/ex_21/create_labeled_dataset.py:163]: I think there was a change in the way that file names were formatted, so this is a fix to make sure this function can handle the new and old filename format.

File: src/mozilla_sec_eia/models/sec10k/ex_21/inference.py [hunk 5]
```diff
@@ -382,10 +374,10 @@ def postprocess(self, all_outputs):
         """Return logits, model predictions, and the extracted dataframe."""
         logits = all_outputs["logits"]
         predictions = all_outputs["logits"].argmax(-1).squeeze().tolist()
-        output_df = self.extract_table(all_outputs)
+        output_df = self._extract_table(all_outputs)
         return logits, predictions, output_df
 
-    def extract_table(self, all_outputs):
+    def _extract_table(self, all_outputs):
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/ex_21/inference.py:380]: I was getting confused between this `extract_table` function that's within the LayoutLM pipeline and the `extract_filings` function which orchestrates running the model and extracting a table so I stuck an underscore onto this one.

File: src/mozilla_sec_eia/models/sec10k/utils/cloud.py [hunk 6]
```diff
@@ -62,6 +62,12 @@ def from_10k(cls, filename: str, sec10k_text: str, ex_21_version: str):
 
     def save_as_pdf(self, file: BinaryIO):
         """Save Exhibit 21 as a PDF in `file`, which can be in memory or on disk."""
+        # TODO: probably should make a "corrections" file that has CSS/HTML replacements
+        # to make PDF render
+        if "border-bottom: black thin solid;" in self.ex_21_text:
+            self.ex_21_text = self.ex_21_text.replace(
+                "border-bottom: black thin solid;", "border-bottom: 1px solid black;"
+            )
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/utils/cloud.py:70]: this is a super janky fix to make one of the Ex.21 PDFs render. I imagine that we will probably want to have a separate cleaning function with a dictionary of common CSS/HTML replacements that make the Ex. 21 PDF render.

File: src/mozilla_sec_eia/models/sec10k/ex_21/rename_labeled_filings.py [hunk 7]
```diff
@@ -24,13 +24,17 @@ def rename_filings():
     archive = GCSArchive()
     bucket = archive._labels_bucket
 
-    labeled_bucket_name = "labeled/"
+    labeled_bucket_name = labeled_bucket_name
```
Comment (katie-lamb) [inline src/mozilla_sec_eia/models/sec10k/ex_21/rename_labeled_filings.py:27]: oops i'll get rid of this

File: src/mozilla_sec_eia/library/validation_helpers.py [hunk 8]
```diff
@@ -1,9 +1,14 @@
 """Implement common utilities/functions for validating models."""
 
+import json
+import re
 from importlib import resources
 
 import pandas as pd
 
+from mozilla_sec_eia.models.sec10k.ex_21.inference import clean_extracted_df
```
Comment (zschira) [inline src/mozilla_sec_eia/library/validation_helpers.py:9]: As we move towards this becoming a `pudl_models` repo, I think it would be best to keep everything in the `library` package totally generic and avoid referencing any specific models. Keeping all the models isolated in their own packages will make the code structure cleaner as the repo grows, and eventually allow us to run models in their own isolated environments so we don't have to deal with dependency conflicts between models.

I think `strip_down_company_names` and `fill_nulls_for_comparison` are pretty generic and can be left here, but `clean_ex21_validation_set` would probably be better moved into `src/mozilla_sec_eia/models/sec10k/ex_21` somewhere.

File: src/mozilla_sec_eia/models/sec10k/ex_21/create_labeled_dataset.py (hunk 4 above)
Comment (zschira) [reply src/mozilla_sec_eia/models/sec10k/ex_21/create_labeled_dataset.py:163]: Not a recommendation for this PR, but at some point I'd like to get the paths/filename handling ironed out and made more consistent throughout. It's gotten a bit janky, which I think makes sense when we've been rapidly prototyping and iterating in parallel, but would be nice to leave it a little more consistent so we don't hate ourselves in the future haha

File: src/mozilla_sec_eia/models/sec10k/utils/cloud.py (hunk 6 above)
Comment (zschira) [reply src/mozilla_sec_eia/models/sec10k/utils/cloud.py:70]: Yeah there's a bunch of cases in the full extraction where things didn't render properly. I might be able to add some better logging to capture these better in the future.

**Review by zschira — CHANGES_REQUESTED** [review]
One change requested on the code structure, but otherwise looks good!

**Review by zschira — APPROVED** [review]
Looks good!
